src/client.rs

```diff
@@ -538,6 +534,28 @@ where
                 continue;
             }
 
+            // Get a pool instance referenced by the most up-to-date
+            // pointer. This ensures we always read the latest config
+            // when starting a query.
+            let mut pool =
+                match get_pool(self.target_pool_name.clone(), self.target_user_name.clone()) {
+                    Some(pool) => pool,
+                    None => {
+                        error_response(
+                            &mut self.write,
+                            &format!(
+                                "No pool configured for database: {:?}, user: {:?}",
+                                self.target_pool_name, self.target_user_name
+                            ),
+                        )
+                        .await?;
+                        return Err(Error::ClientError);
+                    }
+                };
+
+            // The query router determines where the query is going to go,
+            // e.g. primary, replica, which shard.
+            let mut query_router = QueryRouter::new(pool.clone());
```

Comment:
The query router should be outside the client loop because it keeps the state of the client connection, re: current shard, current role, etc. For example, when a client issues `SET SHARD TO`, it expects that setting to persist until it's changed again by the client -- it must survive multiple transactions.

The `QueryRouter` just needs to fetch the pool settings dynamically like it used to before (you can pass it the `target_pool_name` and `target_user_name`.